doc/SAI_Proposal_Generic_Extensions.md

```diff
@@ -0,0 +1,423 @@
+
+![](sai_p4.png)
+
+## An Example - L4LB ##
+The API usage is explained via an example. The example chosen is L4 Load Balancer. The following steps summarize the operation of the HW block.
+
+![](lb_arch.png)
+
+
+- Hybrid architecture where the initial load-balancing decision is taken at the switch and connection tracking is implemented at the edge (NIC/Appliance etc)
+- Switch will advertise VIP through routing protocols in order to attract initial packet(s) to the switch.
+- Once the switch receives the first packet (SYN), it will choose one of the backends from a pool, encapsulate the packet with the appropriate tunnel header and route the packet based on destination IP in the tunnel header.
+- Switch will also send a copy (ICMP response) of the packet to the source VM so a flow entry can be installed there.
+- Once a flow has been established, subsequent packets of the flow will be sent directly from the source VM to the destination VM.
+
+From an application standpoint, 
+- It will push a VIP and the corresponding DIP pool to the DUT modeled below by the VIP table in the red block. 
+- Use an ACL table to filter flows that are allowed to be load-balanced 
+- The encap information can be provided by either passing in DIP list directly and leave the nexthop resolution to the SAI implementation or by explicitly passing the Tunnel nexthop IDs as the DIP pool.
+
+![](l4lb.png)
+
+## JSON data ##
+The following example is the actual data that flows down the software stack. Note that this format adheres to the grammar specified in the definition of sai_json_t.
+```json
+"attributes": [
+    {
+        "virtual_ip": {
+            "sai_metadata": {
+                "sai_attr_value_type": "SAI_ATTR_VALUE_TYPE_IP_ADDRESS"
+            },
+            "value": "10.1.1.1"
+        }
+    },
+    {
+        "nexthop_list": {
+            "sai_metadata": {
+                "sai_attr_value_type": "SAI_ATTR_VALUE_TYPE_OBJECT_LIST"
+            },
+            "value": [ "0x1500000000001fa", "0x1500000000001fb" ]
+        }
+    }
+]
+```
+
+## SAI ACL considerations ##
+The new VIP table is conditionally evaluated based on policies in the ACL tables. This requires a new action type in saiacl.h.
+SAI ACL action type is today exposed via 2 mechanisms. * sai_acl_action_type_t - This type is only used today to query the SAI ACL action capabilities * SAI_ACL_ENTRY_ATTR_ACTION_START/END - An attribute specifiying the ACL action type for an entry
+While there are multiple mechanisms to extend these, one recommendation is as below. 
+- Extend sai_acl_acion_type_t with a custom range just like all the attribute enums for SAI objects 
+  - SAI_ACL_ACTION_TYPE_CUSTOM_RANGE 
+  - SAI_ACL_ACTION_TYPE_DO_LB = SAI_ACL_ACTION_TYPE_CUSTOM_RANGE + 1 
```

Comment:
This requires changing the SAI header, thereby defeating the state purpose of avoding header file changes/recompilation.

Reply:
I agree. The design of this particular application required changing the ACL object to get it to work. You will agree though that programming the vip_table here does not require any changes to SAI? That is ultimately the goal to allow programming of niche H/W blocks without modifying SAI.